v2/datastream-to-spanner/terraform/samples/all-infrastructure/README.md

```diff
@@ -0,0 +1,554 @@
+of the session file to the `var.dataflow_params.template_params.sessionFilePath`
+variable.
+
+> **_NOTE:_** At the time of generating the session file via SMT, the session
+> file is uploaded to a user-specified (or auto-generated) bucket, so you don't
+> need to upload the session file to GCS manually.
+
+### Cross project writes to Spanner
+
+The dataflow job can write to Spanner in a different project. In order to do so,
+the service account running the Dataflow job needs to have the
+`roles/spanner.databaseAdmin`role (or the corresponding permissions to write
+data to Spanner).
+
+After adding these permissions, configure the
+`var.dataflow_params.template_params.projectId` variable.
+
+### Adding access to Terraform service account
+
+#### Using custom role and granular permissions (recommended)
+
+You can run the following gcloud command to create a custom role in your GCP
+project. 
+
+```shell
+gcloud iam roles create live_migrations_role --project=<YOUR-PROJECT-ID> --file=perms.yaml --quiet
+```
+
+The `YAML` file required for the above will be like so - 
+
+```shell
+title: "Live Migrations Custom Role"
+description: "Custom role for Spanner live migrations."
+stage: "GA"
+includedPermissions:
+- iam.roles.get
+- iam.serviceAccounts.actAs
+- datastream.connectionProfiles.create
+....add all permissions from the list defined above.
+```
+
+Then attach the role to the service account - 
+
+```shell
+gcloud iam service-accounts add-iam-policy-binding <YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com \
+    --member=<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com --role=projects/<YOUR-PROJECT-ID>/roles/live_migrations_role \
+    --condition=CONDITION
+```
+
+#### Using pre-defined roles
+
+You can run the following shell script to add roles to the service account
+being used to run Terraform. This will have to done by a user which has the
+authority to grant the specified roles to a service account -
+
+```shell
+#!/bin/bash
+
+# Service account to be granted roles
+SERVICE_ACCOUNT="<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com"
+
+# Project ID where roles will be granted
+PROJECT_ID="<YOUR-PROJECT-ID>"
+
+# Array of roles to grant
+ROLES=(
+  "roles/<role1>"
+  "roles/<role2>"
+)
+
+# Loop through each role and grant it to the service account
+for ROLE in "${ROLES[@]}"
+do
+  gcloud projects add-iam-policy-binding "$PROJECT_ID" \
+    --member="serviceAccount:${SERVICE_ACCOUNT}" \
+    --role="$ROLE"
+done
+```
+
+### Verifying access in the Terraform service account
+
+#### Using custom role and granular permissions (recommended)
+
+Verify that the custom role is attached to the service account - 
+
+```shell
+gcloud projects get-iam-policy <YOUR-PROJECT-ID>  \
+--flatten="bindings[].members" \
+--format='table(bindings.role)' \
+--filter="bindings.members:<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com"
+```
+
+Verify that the role has the correct set of permissions
```

Comment:
Is the verification manual ? Can it be automated ?

Reply:
These are pretty sensitive permissions and should not be automated, in general. That is why we provide the recipes to the user when dealing with Terraform permissions themselves. This is a one-time setup generally done by someone with elevated privileges.

Can probably be automated by writing a shell script over gCloud I guess. Will leave that to the user for now and see if it's a consistent concern and we need to do something about it. 

Overall, the feedback is pretty consistent from terraform side in my investigations.